Related tags: build html server-side, eliminate cookies

evazion opened the pull request:
Related to #3955. This refactors the related tags code so that the related tags html is built server-side instead of client-side.

Instead of calling `/related_tags.json` and building the html in javascript, we call `/related_tags.js` and generate the html server-side. This allows a lot of the related tags javascript to be eliminated.

This also allows the `favorite_tags` and `recent_tags` cookies to be eliminated. The recent tags are now taken as the top 20 tags used by the current user in the last 20 edits made within the last hour. Previously it was just the full set of tags from their most recently edited post.

evazion commented:
Working on fixing those things and some other issues.

evazion commented:
Fixed the issues mentioned above.

This is the overall logic for how this works:

* When the page first loads, the `#related-tags-container` begins with empty `<div>`'s for each column.

* When the related tags are first shown, it triggers a call to `/related_tags.js?user_tags=true` to fill in the user's recent and frequent tags, and a call to `/source.js` to fill in the translated tag and artist tag columns. These calls are made only once, the first time the related tags section is shown.

* When the `Related tags` button is clicked, it triggers a call to `/related_tags.js?query=$query` to fill in the main related tags and the wiki tags.

This fixes a couple issues with the current implementation of related tags. The first is that `/source.js` isn't called when you open the tag edit dialog with shift+e, so the artist and translated tags aren't listed there. The second is that `/related_tags.json` and `/source.js` are called each time you switch to the "Edit" tab, so it can be called multiple times if you switch back and forth.

BrokenEagle commented on app/javascript/src/javascripts/related_tag.js:
```diff
@@ -0,0 +1,142 @@
+
+  // Initialize the recent/favorite/translated/artist tag columns once, the first time the related tags are shown.
+  $(document).one("danbooru:show-related-tags", RelatedTag.initialize_recent_and_favorite_tags);
+  $(document).one("danbooru:show-related-tags", Uploads.fetch_data_manual);
```
Shouldn't it be **on** instead of **one**?

```javascript
 $(document).on("danbooru:show-related-tags", RelatedTag.initialize_recent_and_favorite_tags);
 $(document).on("danbooru:show-related-tags", Uploads.fetch_data_manual);
```

evazion commented on app/views/posts/partials/show/_edit.html.erb:
```diff
@@ -82,20 +82,12 @@
-    <div id="related-tags" class="related-tags">
-    </div>
-  </div>
+  <%= render "related_tags/container", related_tags: nil %>
```
This was code left behind by a previous refactoring.

evazion commented:
Fixed the above issues in 6d293df.

Returning false in `RelatedTag.show` has the same effect as calling `event.preventDefault()` ([ref]). I did it this way because sometimes `RelatedTag.show` is called as an event handler (when `#show-related-tags-link` is clicked), in which case we need to call `event.preventDefault()` , but other times `RelatedTag.show()` is called directly, without an event param, so we need to check that `event` isn't null for that case.

Returning false avoids the null check, but you're right that what's going on here isn't very clear, so I removed it.

[ref]: https://css-tricks.com/return-false-and-prevent-default/
